src/utilities/new-sequence-editor/seq-json-linter.ts

```diff
@@ -0,0 +1,89 @@
+import { linter, type Diagnostic } from '@codemirror/lint';
+import type { Extension, Text } from '@codemirror/state';
+import type { CommandDictionary } from '@nasa-jpl/aerie-ampcs';
+// @ts-expect-error library does not include type declarations
+import { parse as jsonSourceMapParse } from 'json-source-map';
+
+type JsonSourceMapPointerPosition = {
+  column: number;
+  line: number;
+  pos: number;
+};
+
+type JsonSourceMapPointer = {
+  key: JsonSourceMapPointerPosition;
+  keyEnd: JsonSourceMapPointerPosition;
+  value: JsonSourceMapPointerPosition;
+  valueEnd: JsonSourceMapPointerPosition;
+};
+
+/**
+ * Helper for getting an error position of JSON.prase throws a SyntaxError.
+ * @see https://github.com/codemirror/lang-json/blob/main/src/lint.ts
+ */
+function getErrorPosition(error: SyntaxError, doc: Text): number {
+  let m;
+
+  if ((m = error.message.match(/at position (\d+)/))) {
+    return Math.min(+m[1], doc.length);
+  }
+
+  if ((m = error.message.match(/at line (\d+) column (\d+)/))) {
+    return Math.min(doc.line(+m[1]).from + +m[2] - 1, doc.length);
+  }
+
+  return 0;
+}
+
+/**
+ * Linter function that returns a Code Mirror extension function.
+ * Can be optionally called with a command dictionary so it's available during linting.
+ */
+export function seqJsonLinter(commandDictionary: CommandDictionary | null = null): Extension {
+  return linter(view => {
+    const diagnostics: Diagnostic[] = [];
+
+    try {
+      const text = view.state.doc.toString();
+      const sourceMap = jsonSourceMapParse(text);
+
+      if (commandDictionary) {
+        for (const [key, pointer] of Object.entries<JsonSourceMapPointer>(sourceMap.pointers)) {
+          const stemMatch = key.match(/\/steps\/\d+\/stem/);
+
+          if (stemMatch) {
+            const stemValue = view.state.doc.sliceString(pointer.value.pos, pointer.valueEnd.pos);
```

Comment:
This looks like it's going through a JSON object, looking for keys that match any string containing `steps#stem`. Then it continues to use the value at that key. Is `json-source-map` necessary for this? Is it not possible to just `JSON.parse` the text to get the actual JSON object which can then be traversed to get the values instead of using string positions to extract from the `view.state.doc` after parsing?

Reply:
I'll defer to @joswig on this one. I think it could be accomplished using the native `JSON.parse` like you suggested but `json-source-map` provides pointers to deeply nested properties that we'd need to traverse to find if we went with that route?

Reply:
We use the string positions to create the code mirror diagnostics in the correct position